### PR TITLE
Updating ssc_interface to be platform independant.

### PR DESCRIPTION
**Pull request by as-snehagn**
## Status
**PRODUCTION / DEVELOPMENT**
** Tested on PACMod Lexus **

## Description
Replacing DBW specific messages with SSC outputs and minor updates to associated computations to make ssc_interface platform independent.
Replacing velocity_accel with velocity_accel_cov message.

## Todos
- [x] Tested on the New Eagle Pacifica
- [x] Tested on the Dataspeed Fusion

### CODE REVIEW

File: ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.h [hunk 1]
```diff
@@ -118,13 +119,12 @@ class SSCInterface
   void callbackFromVehicleCmd(const autoware_msgs::VehicleCmdConstPtr& msg);
   void callbackFromEngage(const std_msgs::BoolConstPtr& msg);
   void callbackFromSSCModuleStates(const automotive_navigation_msgs::ModuleStateConstPtr& msg);
-  void callbackFromSSCFeedbacks(const automotive_platform_msgs::VelocityAccelConstPtr& msg_velocity,
```
Comment (JWhitleyAStuff) [inline ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.h]: Why was this changed?

File: ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.cpp [hunk 2]
```diff
@@ -19,7 +19,7 @@
 SSCInterface::SSCInterface() : nh_(), private_nh_("~"), engage_(false), command_initialized_(false)
 {
   // setup parameters
-  private_nh_.param<bool>("use_rear_wheel_speed", use_rear_wheel_speed_, true);
+  // private_nh_.param<bool>("use_rear_wheel_speed", use_rear_wheel_speed_, true); #Redudant, being estimated in SSC
```
Comment (JWhitleyAStuff) [inline ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.cpp]: Please remove these items instead of commenting them out. We can always roll back in the commit history if we need to see what was here.

File: ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.cpp [hunk 3]
```diff
@@ -50,17 +50,17 @@ SSCInterface::SSCInterface() : nh_(), private_nh_("~"), engage_(false), command_
...
       new message_filters::Subscriber<automotive_platform_msgs::GearFeedback>(nh_, "as/gear_feedback", 10);
   velocity_accel_sub_ =
-      new message_filters::Subscriber<automotive_platform_msgs::VelocityAccel>(nh_, "as/velocity_accel", 10);
-  // subscribers from PACMod
-  wheel_speed_sub_ =
-      new message_filters::Subscriber<pacmod_msgs::WheelSpeedRpt>(nh_, "pacmod/parsed_tx/wheel_speed_rpt", 10);
+      new message_filters::Subscriber<automotive_platform_msgs::VelocityAccelCov>(nh_, "as/velocity_accel_cov", 10); 
+  //VelocityAccel message will be deprecated
+  //wheel_speed_sub_ =
+  //    new message_filters::Subscriber<pacmod_msgs::WheelSpeedRpt>(nh_, "pacmod/parsed_tx/wheel_speed_rpt", 10);
```
Comment (JWhitleyAStuff) [inline ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.cpp]: Same here. Remove instead of commenting.

File: ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.cpp [hunk 4]
```diff
@@ -107,36 +107,35 @@ void SSCInterface::callbackFromSSCModuleStates(const automotive_navigation_msgs:
...
-  // current speed
-  double speed =
-      !use_rear_wheel_speed_ ?
-          (msg_velocity->velocity) :
-          (msg_wheel_speed->rear_left_wheel_speed + msg_wheel_speed->rear_right_wheel_speed) * tire_radius_ / 2.;
+  // current speed - implementated inside SSC for estimating velocityaccelcov message.
+  //double speed =
+  //    !use_rear_wheel_speed_ ?
+  //       (msg_velocity->velocity) :
+  //        (msg_wheel_speed->rear_left_wheel_speed + msg_wheel_speed->rear_right_wheel_speed) * tire_radius_ / 2.;
```
Comment (JWhitleyAStuff) [inline ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.cpp]: Here too.

**Review by JWhitleyWork — COMMENTED** [review]
See comments.

**Comment by JWhitleyAStuff** [comment]
Also, Github Notifications is currently experiencing an outage so CI was not kicked off for this PR. Once you fix the issues I mentioned above, I may have you kill and re-initialize this PR to kick off CI. See https://status.circleci.com/ for status.

File: ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.h (hunk 1 above)
Comment (as-snehagn) [reply ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.h]: The velocity_accel and velocity_accel_cov essentially contain the same information. We are deprecating velocity_accel and will be switching to only using the velocity_accel_cov message, as @Daniel-Stanek mentioned to me.

Comment (JWhitleyAStuff) [reply ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.h]: Have you verified with Dan that this is what we are using on any vehicles with a _released_ version of SSC?

File: ros/src/actuation/vehicles/packages/as/launch/ssc_interface.launch [hunk 5]
```diff
@@ -12,7 +11,6 @@
 	<arg name="max_curvature_rate" default="0.75"/>
 
 	<node pkg="as" type="ssc_interface" name="ssc_interface" output="screen">
-		<param name="use_rear_wheel_speed" value="$(arg use_rear_wheel_speed)" />
```
Comment (icolwell-as) [inline ros/src/actuation/vehicles/packages/as/launch/ssc_interface.launch]: Can you provide some explanation as to why this was removed?

File: ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.cpp [hunk 6]
```diff
@@ -107,36 +103,29 @@ void SSCInterface::callbackFromSSCModuleStates(const automotive_navigation_msgs:
...
                                             const automotive_platform_msgs::BrakeFeedbackConstPtr& msg_brake,
                                             const automotive_platform_msgs::GearFeedbackConstPtr& msg_gear,
-                                            const pacmod_msgs::WheelSpeedRptConstPtr& msg_wheel_speed,
-                                            const pacmod_msgs::SystemRptFloatConstPtr& msg_steering_wheel)
+                                            const automotive_platform_msgs::SteeringFeedbackConstPtr& msg_steering_wheel)
 {
   ros::Time stamp = msg_velocity->header.stamp;
-
-  // current speed
-  double speed =
```
Comment (icolwell-as) [inline ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.cpp]: Probably would have been cleaner to just change this line to `double speed = msg_velocity->velocity`, then the rest of the changes aren't needed.

**Review by icolwell-as — COMMENTED** [review]
Can you explain how this PR makes the ssc_interface platform-independant? It seems like all you are doing is changing some message types and the way speed is calculated/used.

Also, we should probably be making this PR into autoware master, otherwise we'll be dealing with conflicts later if someone else decides to edit ssc_interface on the autoware side.

File: ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.h [hunk 7]
```diff
@@ -92,7 +91,7 @@ class SSCInterface
   double deceleration_limit_;  // [m/s^2]
   double max_curvature_rate_;  // [rad/m/s]
 
-  bool use_rear_wheel_speed_;     // instead of 'as/velocity_accel'
+  //bool use_rear_wheel_speed_;     // instead of 'as/velocity_accel'
```
Comment (icolwell-as) [inline ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.h]: Delete commented code.

**Review by icolwell-as — CHANGES_REQUESTED** [review]

File: ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.h (hunk 1 above)
Comment (JWhitleyAStuff) [reply ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.h]: It looks like 2.1.1 release of `ros_vehicle` (which means all ssc_ packages) provide this. I'm good with it.

File: ros/src/actuation/vehicles/packages/as/launch/ssc_interface.launch (hunk 5 above)
Comment (as-snehagn) [reply ros/src/actuation/vehicles/packages/as/launch/ssc_interface.launch]: SSC is used in estimating the current speed of the vehicle and it uses the wheel speed, this is redundant in here. I have removed this usage in code.

**Comment by JWhitleyAStuff** [comment]
@icolwell-as - We want to test this on all of our own platforms that we have access to before merging it to autoware/master. Also, it won't get merged there until 1.13 (feature freeze for 1.12 is on now, I believe). The cycle would be too long to get it back here.

In answer to how it makes it "platform-independent," it removes a specific dependency on `pacmod_msgs` which is tied to PACMod-only vehicles. This change makes it only SSC-dependent which is more abstract and can control 3 DBW systems.

File: ros/src/actuation/vehicles/packages/as/launch/ssc_interface.launch (hunk 5 above)
Comment (icolwell-as) [reply ros/src/actuation/vehicles/packages/as/launch/ssc_interface.launch]: Thanks! It looks like the SSC speed signal is much cleaner too
Velocity Signal

**Comment by icolwell-as** [comment]
Thanks for the clarification Josh! I didn't know the SSC can work with anything other than PACMod.
Makes perfect sense.

**Review by Daniel-Stanek — APPROVED** [review]

**Review by JWhitleyWork — APPROVED** [review]

**Comment by icolwell-as** [comment]
PRs shouldn't be merged while somebody still has open change requests. We just merged in leftover commented code to master:
https://github.com/astuff/autoware/commit/10e79d903f96ab48162138547d57ee176716dc7b#diff-ee9028409296ee6e4c01fea10f9f0c0cR94

**Comment by JWhitleyWork** [comment]
@icolwell-as - I apologize. I'll try to watch more carefully in the future and I'll fix this one.
